model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorage.java

```diff
@@ -152,29 +194,91 @@ private static String toOrderString(QueryParameters.OrderBy<?> orderBy) {
 
     @Override
     public Stream<V> read(QueryParameters<M> queryParameters) {
+        if (LockObjectsForModification.isEnabled(session, storedEntityDescriptor.getModelTypeClass())) {
+            return pessimisticQueryRead(queryParameters);
+        }
+
+        Query<E> query = prepareQueryWithPrefixAndParameters(null, queryParameters);
+        CloseableIterator<E> iterator = paginateQuery(query, queryParameters.getOffset(),
+                queryParameters.getLimit()).iterator();
+        return closing(StreamSupport.stream(Spliterators.spliteratorUnknownSize(iterator, 0), false))
+                .onClose(iterator::close)
+                .filter(Objects::nonNull) // see https://github.com/keycloak/keycloak/issues/9271
+                .map(this.delegateProducer);
+    }
+
+    private Stream<V> pessimisticQueryRead(QueryParameters<M> queryParameters) {
+        DefaultModelCriteria<M> dmc = queryParameters.getModelCriteriaBuilder();
+        List<String> idsToRead = new LinkedList<>();
+
+        dmc.partiallyEvaluate((field, operator, operatorArguments) -> {
```

Comment:
What if there is a query (maybe in future) that contains EQ|IN operator for ID but also something else i.e `... &&  timestamp < xyz`?

Reply:
The rest of the criteria is tested after entity is loaded to Keycloak here: https://github.com/keycloak/keycloak/pull/16619/files#diff-2cd58c3c33d4f93a5ff9b8262ffe66ebc3f375e1c7a3d519652a8e293d50bf5dR240